pkg/dataset/downsample.go

```diff
@@ -0,0 +1,76 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package dataset
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/rollup"
+)
+
+const (
+	defaultDownsampleState      = true
+	defaultDownsampleResolution = "short:5m:90d,long:1h:395d" // label:resolution:retention
+)
+
+var (
+	setDefaultDownsampleStateSQL      = "SELECT prom_api.set_automatic_downsample($1)"
+	setDefaultDownsampleResolutionSQL = "SELECT prom_api.set_downsample_resolution($1)"
+
+	defaultDownsampleStateVar = defaultDownsampleState
+)
+
+type Downsample struct {
+	Automatic  *bool  `yaml:"automatic,omitempty"`
+	Resolution string `yaml:"resolution,omitempty"`
+}
+
+func (d *Downsample) Apply(conn *pgx.Conn) error {
+	d.applyDefaults()
+
+	queries := map[string]interface{}{
+		setDefaultDownsampleStateSQL: d.Automatic,
+	}
+	if *d.Automatic {
+		if err := handleDownsampling(conn, d.Resolution); err != nil {
+			return fmt.Errorf("handle downsample: %w", err)
+		}
+		queries[setDefaultDownsampleResolutionSQL] = d.Resolution
+		log.Info("msg", fmt.Sprintf("Setting metric downsample resolution to %s", d.Resolution))
+	}
+	log.Info("msg", fmt.Sprintf("Setting metric automatic downsample to %t", *d.Automatic))
+
+	for sql, param := range queries {
+		if _, err := conn.Exec(context.Background(), sql, param); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Downsample) applyDefaults() {
+	if d.Automatic == nil {
```

Comment:
Do we validate `automatic: false` and no resolution specified?

Reply:
If no resolution is specified (which is the default case), we consider it `5m` & `1h` downsampling case. If a user does not want downsampling, then has to set `automatic: false`.